File: Freestylecoding.Math.FSharp.Tests/Natural/ComparisonOperatorsNaturalNaturalBool.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public ComparisonOperatorsNaturalNaturalBool() =
    static member private op_GT<'T when 'T :> System.Numerics.IComparisonOperators<'T,'T,bool>>( left:'T, right:'T ) =
        'T.op_GreaterThan( left, right )
    static member private op_GTE<'T when 'T :> System.Numerics.IComparisonOperators<'T,'T,bool>>( left:'T, right:'T ) =
        'T.op_GreaterThanOrEqual( left, right )
    static member private op_LT<'T when 'T :> System.Numerics.IComparisonOperators<'T,'T,bool>>( left:'T, right:'T ) =
        'T.op_LessThan( left, right )
    static member private op_LTE<'T when 'T :> System.Numerics.IComparisonOperators<'T,'T,bool>>( left:'T, right:'T ) =
        'T.op_LessThanOrEqual( left, right )

    [<Theory>]
    [<InlineData( 0u, 1u, false )>]
    [<InlineData( 1u, 0u, true )>]
    [<InlineData( 1u, 1u, false )>]
    member public this.GT_Sanity left right expected =
        Assert.Equal(
            expected,
            ComparisonOperatorsNaturalNaturalBool.op_GT(
                Natural( [left] ),
                Natural( [right] )
            )
        )

    [<Fact>]
    member public this.GT_BiggerLeft () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_GT(
                Natural( [0xBADu; 0xDEADBEEFu] ),
                Natural( [0xDEADBEEFu] )
            )
        )

    [<Fact>]
    member public this.GT_BiggerRight () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_GT(
                Natural( [0xDEADBEEFu] ),
                Natural( [0xBADu; 0xDEADBEEFu] )
            )
        )
    
    [<Fact>]
    member public this.GT_CascadeGreaterThan () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_GT(
                Natural( [1u; 1u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.GT_CascadeEqual () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_GT(
                Natural( [1u; 0u] ),
                Natural( [1u; 0u] )
            )
        )    

    [<Fact>]
    member public this.GT_CascadeLessThan () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_GT(
                Natural( [1u; 0u] ),
                Natural( [1u; 1u] )
            )
        )

    [<Theory>]
    [<InlineData( 0u, 1u, true )>]
    [<InlineData( 1u, 0u, false )>]
    [<InlineData( 1u, 1u, false )>]
    member public this.LT_Sanity left right expected =
        Assert.Equal(
            expected,
            ComparisonOperatorsNaturalNaturalBool.op_LT(
                Natural( [left] ),
                Natural( [right] )
            )
        )

    [<Fact>]
    member public this.LT_BiggerLeft () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_LT(
                Natural( [0xBADu; 0xDEADBEEFu] ),
                Natural( [0xDEADBEEFu] )
            )
        )

    [<Fact>]
    member public this.LT_BiggerRight () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_LT(
                Natural( [0xDEADBEEFu] ),
                Natural( [0xBADu; 0xDEADBEEFu] )
            )
        )
    
    [<Fact>]
    member public this.LT_CascadeGreaterThan () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_LT(
                Natural( [1u; 1u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.LT_CascadeEqual () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_LT(
                Natural( [1u; 0u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.LT_CascadeLessThan () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_LT(
                Natural( [1u; 0u] ),
                Natural( [1u; 1u] )
            )
        )

    [<Theory>]
    [<InlineData( 0u, 1u, false )>]
    [<InlineData( 1u, 0u, true )>]
    [<InlineData( 1u, 1u, true )>]
    member public this.GTE_Sanity left right expected =
        Assert.Equal(
            expected,
            ComparisonOperatorsNaturalNaturalBool.op_GTE(
                Natural( [left] ),
                Natural( [right] )
            )
        )

    [<Fact>]
    member public this.GTE_BiggerLeft () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_GTE(
                Natural( [0xBADu; 0xDEADBEEFu] ),
                Natural( [0xDEADBEEFu] )
            )
        )

    [<Fact>]
    member public this.GTE_BiggerRight () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_GTE(
                Natural( [0xDEADBEEFu] ),
                Natural( [0xBADu; 0xDEADBEEFu] )
            )
        )
    
    [<Fact>]
    member public this.GTE_CascadeGreaterThan () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_GTE(
                Natural( [1u; 1u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.GTE_CascadeEqual () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_GTE(
                Natural( [1u; 0u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.GTE_CascadeLessThan () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_GTE(
                Natural( [1u; 0u] ),
                Natural( [1u; 1u] )
            )
        )

    [<Theory>]
    [<InlineData( 0u, 1u, true )>]
    [<InlineData( 1u, 0u, false )>]
    [<InlineData( 1u, 1u, true )>]
    member public this.LTE_Sanity left right expected =
        Assert.Equal(
            expected,
            ComparisonOperatorsNaturalNaturalBool.op_LTE(
                Natural( [left] ),
                Natural( [right] )
            )
        )

    [<Fact>]
    member public this.LTE_BiggerLeft () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_LTE(
                Natural( [0xBADu; 0xDEADBEEFu] ),
                Natural( [0xDEADBEEFu] )
            )
        )

    [<Fact>]
    member public this.LTE_BiggerRight () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_LTE(
                Natural( [0xDEADBEEFu] ),
                Natural( [0xBADu; 0xDEADBEEFu] )
            )
        )
    
    [<Fact>]
    member public this.LTE_CascadeGreaterThan () =
        Assert.False(
            ComparisonOperatorsNaturalNaturalBool.op_LTE(
                Natural( [1u; 1u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.LTE_CascadeEqual () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_LTE(
                Natural( [1u; 0u] ),
                Natural( [1u; 0u] )
            )
        )
    
    [<Fact>]
    member public this.LTE_CascadeLessThan () =
        Assert.True(
            ComparisonOperatorsNaturalNaturalBool.op_LTE(
                Natural( [1u; 0u] ),
                Natural( [1u; 1u] )
            )
        )

File: Freestylecoding.Math.CSharp.Tests/Natural/ComparisonOperatorsNaturalNaturalBool.cs
using System.Numerics;
using Xunit;

namespace Freestylecoding.Math.CSharp.Tests.Natural;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class ComparisonOperatorsNaturalNaturalBool {
	private static bool op_GT<T>( T left, T right ) where T : IComparisonOperators<T, T, bool> =>
		left > right;
	private static bool op_GTE<T>( T left, T right ) where T : IComparisonOperators<T, T, bool> =>
		left >= right;
	private static bool op_LT<T>( T left, T right ) where T : IComparisonOperators<T, T, bool> =>
		left < right;
	private static bool op_LTE<T>( T left, T right ) where T : IComparisonOperators<T, T, bool> =>
		left <= right;

	[Theory]
	[InlineData( 0u, 1u, false )]
	[InlineData( 1u, 0u, true )]
	[InlineData( 1u, 1u, false )]
	public void GT_Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			op_GT(
				new Natural( left ),
				new Natural( right )
			)
		);

	[Fact]
	public void GT_BiggerLeft() =>
		Assert.True(
			op_GT(
				FListNatural( 0xBADu, 0xDEADBEEFu ),
				new Natural( 0xDEADBEEFu )
			)
		);

	[Fact]
	public void GT_BiggerRight() =>
		Assert.False(
			op_GT(
				new Natural( 0xDEADBEEFu ),
				FListNatural( 0xBADu, 0xDEADBEEFu )
			)
		);

	[Fact]
	public void GT_CascadeGreaterThan() =>
		Assert.True(
			op_GT(
				FListNatural( 1u, 1u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void GT_CascadeEqual() =>
		Assert.False(
			op_GT(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void GT_CascadeLessThan() =>
		Assert.False(
			op_GT(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 1u )
			)
		);

	[Theory]
	[InlineData( 0u, 1u, true )]
	[InlineData( 1u, 0u, false )]
	[InlineData( 1u, 1u, false )]
	public void LT_Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			op_LT(
				new Natural( left ),
				new Natural( right )
			)
		);

	[Fact]
	public void LT_BiggerLeft() =>
		Assert.False(
			op_LT(
				FListNatural( 0xBADu, 0xDEADBEEFu ),
				new Natural( 0xDEADBEEFu )
			)
		);

	[Fact]
	public void LT_BiggerRight() =>
		Assert.True(
			op_LT(
				new Natural( 0xDEADBEEFu ),
				FListNatural( 0xBADu, 0xDEADBEEFu )
			)
		);

	[Fact]
	public void LT_CascadeGreaterThan() =>
		Assert.False(
			op_LT(
				FListNatural( 1u, 1u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void LT_CascadeEqual() =>
		Assert.False(
			op_LT(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void LT_CascadeLessThan() =>
		Assert.True(
			op_LT(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 1u )
			)
		);

	[Theory]
	[InlineData( 0u, 1u, false )]
	[InlineData( 1u, 0u, true )]
	[InlineData( 1u, 1u, true )]
	public void GTE_Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			op_GTE(
				new Natural( left ),
				new Natural( right )
			)
		);

	[Fact]
	public void GTE_BiggerLeft() =>
		Assert.True(
			op_GTE(
				FListNatural( 0xBADu, 0xDEADBEEFu ),
				new Natural( 0xDEADBEEFu )
			)
		);

	[Fact]
	public void GTE_BiggerRight() =>
		Assert.False(
			op_GTE(
				new Natural( 0xDEADBEEFu ),
				FListNatural( 0xBADu, 0xDEADBEEFu )
			)
		);

	[Fact]
	public void GTE_CascadeGreaterThan() =>
		Assert.True(
			op_GTE(
				FListNatural( 1u, 1u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void GTE_CascadeEqual() =>
		Assert.True(
			op_GTE(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void GTE_CascadeLessThan() =>
		Assert.False(
			op_GTE(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 1u )
			)
		);

	[Theory]
	[InlineData( 0u, 1u, true )]
	[InlineData( 1u, 0u, false )]
	[InlineData( 1u, 1u, true )]
	public void LTE_Sanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			op_LTE(
				new Natural( left ),
				new Natural( right )
			)
		);

	[Fact]
	public void LTE_BiggerLeft() =>
		Assert.False(
			op_LTE(
				FListNatural( 0xBADu, 0xDEADBEEFu ),
				new Natural( 0xDEADBEEFu )
			)
		);

	[Fact]
	public void LTE_BiggerRight() =>
		Assert.True(
			op_LTE(
				new Natural( 0xDEADBEEFu ),
				FListNatural( 0xBADu, 0xDEADBEEFu )
			)
		);

	[Fact]
	public void LTE_CascadeGreaterThan() =>
		Assert.False(
			op_LTE(
				FListNatural( 1u, 1u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void LTE_CascadeEqual() =>
		Assert.True(
			op_LTE(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 0u )
			)
		);

	[Fact]
	public void LTE_CascadeLessThan() =>
		Assert.True(
			op_LTE(
				FListNatural( 1u, 0u ),
				FListNatural( 1u, 1u )
			)
		);
}